### PR TITLE
Odyssey Stats: replace config api with custom implementation

### PR DESCRIPTION
**Pull request by kangzj**
Related to #75257 

## Proposed Changes

The PR proposes to replace the `@automattic/calypso-config` with `src/lib/config-api` which,

 * Supports custom config data source, which is useful for Odyssey Apps.
 * Restores URL/Hash based feature flag gating for Odyssey Apps.
 * Must be initialized before using.
 * Keeps only one copy of configData.

Also it,
 * Adds unit test ability
 * Copies tests for `@automattic/create-calypso-config` over to make sure our implementation works the same. 
 * Adds automated test on TeamCity

As config is such a fundamental package, and is used everywhere, that I'm not keen to replace it with our implementation even though they are compatible.

Credits go to editing-toolkit for the jest config.

## Testing Instructions

* Ensure Odyssey Stats and Odyssey Widget still work okay for Jetpack bleeding edge
* Ensure Odyssey Stats and Odyssey Widget still work okay for Jetpack from https://github.com/Automattic/jetpack/pull/30147
* Run `cd apps/odyssey-stats && yarn test:js`
* Ensure all tests pass

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: apps/odyssey-stats/src/lib/config-api.js [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	}
+};
+
+configApi.disable = ( feature ) => {
+	if ( configData.features ) {
+		configData.features[ feature ] = false;
+	}
+};
+
+const configOverride = () => {
```
Comment (kangzj) [inline apps/odyssey-stats/src/lib/config-api.js:47]: moved from `load-config.js`

File: apps/odyssey-stats/src/lib/config-api.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
+/**
```
Comment (kangzj) [inline apps/odyssey-stats/src/lib/config-api.js:1]: It appears we could form a real class here

File: apps/odyssey-stats/src/lib/config-api.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * The config package is to provide a comaptible way to access config data exactly like `@automattic/calypso-config`.
```
Comment (jsnmoon) [inline apps/odyssey-stats/src/lib/config-api.js]:
```suggestion
 * The config package is to provide a compatible way to access config data exactly like `@automattic/calypso-config`.
```

File: apps/odyssey-stats/src/lib/config-api.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * - The package supports custom config data source, which is useful for Odyssey Apps.
+ * - Restored feature flag gating for Odyssey Apps.
+ * - The package must be initialized before using with `initConfig`.
```
Comment (jsnmoon) [inline apps/odyssey-stats/src/lib/config-api.js]:
```suggestion
 * - Adds support for custom config data sources for Odyssey Apps.
 * - Adds feature flag gating for Odyssey Apps.
 * - The package must first be initialized before use via an invocation of `initConfig`.
```

File: apps/odyssey-stats/src/lib/create-odyssey-config.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+	// The option enables loading of the whole translation file, and could be optimized by setting it to `true`, which needs the translation chunks in place.
+	// @see https://github.com/Automattic/wp-calypso/blob/trunk/docs/translation-chunks.md
+	productionConfig.features[ 'use-translation-chunks' ] = false;
+
+	// Note: configData is hydrated in https://github.com/Automattic/jetpack/blob/d4d0f987cbf63a864b03b542b7813aabe87e0ed3/projects/packages/stats-admin/src/class-dashboard.php#L214
+	this.configData.features = productionConfig.features;
+};
+
+// Forked from https://github.com/Automattic/wp-calypso/blob/ca7d8fe3e0a5fb87b0659fbab659078ebbfbc7be/packages/calypso-config/src/index.ts#L60
```
Comment (jsnmoon) [inline apps/odyssey-stats/src/lib/create-odyssey-config.js]: Was this forked or copied? If copied, let's say so here. If forked, let's note what we changed.

File: apps/odyssey-stats/src/lib/create-odyssey-config.js [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+				`Could not find config value for key '${ key }'\n` +
+					"Please make sure that if you need it then it has a default value assigned in 'config/_shared.json'"
+			);
+		}
+
+		return undefined;
+	}
+}
+
+// Moved from https://github.com/Automattic/wp-calypso/blob/ca7d8fe3e0a5fb87b0659fbab659078ebbfbc7be/apps/odyssey-stats/src/load-config.js
```
Comment (jsnmoon) [inline apps/odyssey-stats/src/lib/create-odyssey-config.js]:
```suggestion
// Copied from https://github.com/Automattic/wp-calypso/blob/ca7d8fe3e0a5fb87b0659fbab659078ebbfbc7be/apps/odyssey-stats/src/load-config.js
```

**Review by jsnmoon — APPROVED** [review]
These changes look reasonable and work as expected. I left suggestions for improving code documentation inline.

What do you think about converting `config-api.js` and `create-odyssey-config.js` into TypeScript files? Doing so will enforce typing linting across the Odyssey Stats codebase.

File: apps/odyssey-stats/src/lib/create-odyssey-config.js [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+	// The option enables loading of the whole translation file, and could be optimized by setting it to `true`, which needs the translation chunks in place.
+	// @see https://github.com/Automattic/wp-calypso/blob/trunk/docs/translation-chunks.md
+	productionConfig.features[ 'use-translation-chunks' ] = false;
+
+	// Note: configData is hydrated in https://github.com/Automattic/jetpack/blob/d4d0f987cbf63a864b03b542b7813aabe87e0ed3/projects/packages/stats-admin/src/class-dashboard.php#L214
+	this.configData.features = productionConfig.features;
+};
+
+// Forked from https://github.com/Automattic/wp-calypso/blob/ca7d8fe3e0a5fb87b0659fbab659078ebbfbc7be/packages/calypso-config/src/index.ts#L60
```
Comment (kangzj) [inline apps/odyssey-stats/src/lib/create-odyssey-config.js]:
```suggestion
// Copied from https://github.com/Automattic/wp-calypso/blob/ca7d8fe3e0a5fb87b0659fbab659078ebbfbc7be/packages/calypso-config/src/index.ts#L60
```

File: apps/odyssey-stats/src/lib/create-odyssey-config.js (hunk 5 above)
Comment (kangzj) [reply apps/odyssey-stats/src/lib/create-odyssey-config.js]: It's copied. Noted thanks.

**Comment by kangzj** [comment]
> What do you think about converting config-api.js and create-odyssey-config.js into TypeScript files? Doing so will enforce typing linting across the Odyssey Stats codebase.

Good idea! Will add ts support in follow up PRs but probably after this release. Does that sound okay?

File: apps/odyssey-stats/src/lib/test/create-odyssey-config.test.js [hunk 8]
```diff
@@ -0,0 +1,128 @@
+// Forked from https://github.com/Automattic/wp-calypso/blob/6755acc862c0387b7061e13f99fd22c5dd22e458/packages/create-calypso-config/src/test/index.js
```
Comment (kangzj) [inline apps/odyssey-stats/src/lib/test/create-odyssey-config.test.js]:
```suggestion
// Copied from https://github.com/Automattic/wp-calypso/blob/6755acc862c0387b7061e13f99fd22c5dd22e458/packages/create-calypso-config/src/test/index.js
```

**Comment by jsnmoon** [comment]
> Good idea! Will add ts support in follow up PRs but probably after this release. Does that sound okay?

Sounds good to me!

**Review by jsnmoon — APPROVED** [review]

**Pull request merged**